### PR TITLE
Add file contents persistence option to converters.

### PR DESCRIPTION
**Pull request by michaelcfanning**
@lgolding @vinaykapadia @jeff-bacon

### CODE REVIEW

File: src/Sarif/IFileSystem.cs [hunk 1]
```diff
@@ -37,6 +39,17 @@ public interface IFileSystem
...
         /// <summary>
+        /// Opens a binary file, reads all contents into a byte array, and then closes the file.
+        /// </summary>
+        /// <param name="path">
+        /// The file to open for reading.
+        /// </param>
+        /// <returns>
+        /// A byte array containing the contents of the file
+        /// </returns>
+        byte[] ReadAllBytes(string path);
```
Comment (michaelcfanning) [inline src/Sarif/IFileSystem.cs]: What nonsense. We apply this to the interface. Then we apply it to the default concrete implementation. It's already available for direct calls to the actual .NET API. None of these API are particularly abstruse in any case. 

Just a bunch of busy work that consumes vertical space. Bleagh.

File: src/Sarif.UnitTests/Core/FileDataTests.cs [hunk 2]
```diff
@@ -34,7 +33,7 @@ public void FileData_ComputeHashes()
             {
                 File.WriteAllText(filePath, fileContents);
                 FileData fileData = FileData.Create(uri, LoggingOptions.ComputeFileHashes);
-                fileData.Uri.Should().Be(uri.ToString());
+                fileData.Uri.Should().BeNull();
```
Comment (michaelcfanning) [inline src/Sarif.UnitTests/Core/FileDataTests.cs]: >fileData.Uri [](start = 16, length = 12)

file.Uri is actually something like file.NestedUri or file.RelativeUri (for cases when an item is relative to some container file).

File: src/Sarif/IFileSystem.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif/IFileSystem.cs]: MMMV

---
In reply to: [136618018](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136618018) [](ancestors = 136618018)

Comment (jeff-bacon) [reply src/Sarif/IFileSystem.cs]: Is this in reference to the comments or the addition of the new method signature?  

---
In reply to: [136618018](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136618018) [](ancestors = 136618018)

Comment (lgolding) [reply src/Sarif/IFileSystem.cs]: The comments.

---
In reply to: [136618958](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136618958) [](ancestors = 136618958,136618018)

File: SetCurrentVersion.cmd [hunk 3]
```diff
@@ -1,6 +1,6 @@
 set MAJOR_PREVIOUS=1
 set MINOR_PREVIOUS=5
-set PATCH_PREVIOUS=47
+set PATCH_PREVIOUS=46
```
Comment (lgolding) [inline SetCurrentVersion.cmd]: Prev should be 1.6.0, and the new version should be 1.6.1 (really IMO it s.b. 1.7.0 -- this is significant new functionality).

File: src/ConvertToSarif/ConvertOptions.cs [hunk 4]
```diff
@@ -46,5 +46,11 @@ internal class ConvertOptions
             "plugin-assembly-path",
             HelpText = "Path to plugin assembly containing converter types.")]
         public string PluginAssemblyPath { get; internal set; }
+
+
+        [Option(
```
Comment (lgolding) [inline src/ConvertToSarif/ConvertOptions.cs]: >( [](start = 15, length = 1)

Hadn't you already added this option in a previous PR, and didn't you use `'f'` as the short name, per my comment? Between this an my comment on SetCurrentVersion.cmd, I wonder if the branch you're working in is synced to latest changes.

File: src/Sarif/Writers/SarifLogger.cs [hunk 5]
```diff
@@ -157,15 +157,17 @@ private SarifLogger(TextWriter textWriter, LoggingOptions loggingOptions)
             }
         }
 
-        public bool ComputeFileHashes { get { return (_loggingOptions & LoggingOptions.ComputeFileHashes) == LoggingOptions.ComputeFileHashes; } }
+        public bool ComputeFileHashes { get { return _loggingOptions.Includes(LoggingOptions.ComputeFileHashes); } }
```
Comment (lgolding) [inline src/Sarif/Writers/SarifLogger.cs]: >Includes [](start = 69, length = 8)

Yes, this is all from a previous PR as well.

File: src/ConvertToSarif/ConvertOptions.cs (hunk 4 above)
Comment (lgolding) [inline src/ConvertToSarif/ConvertOptions.cs]: >( [](start = 15, length = 1)

Ah, unlike in AnalyzeOptionsBase, `'f'` is already taken here. ok.

File: src/ConvertToSarif/Program.cs [hunk 6]
```diff
@@ -25,18 +26,23 @@ public static int RunConvertFile(ConvertOptions convertOptions)
...
                 if (convertOptions.PrettyPrint)
                 {
-                    toolFormatConversionOptions |= ToolFormatConversionOptions.PrettyPrint;
+                    loggingOptions |= LoggingOptions.PrettyPrint;
                 };
 
                 if (convertOptions.Force)
                 {
-                    toolFormatConversionOptions |= ToolFormatConversionOptions.OverwriteExistingOutputFile;
+                    loggingOptions |= LoggingOptions.OverwriteExistingOutputFile;
```
Comment (lgolding) [inline src/ConvertToSarif/Program.cs]: >OverwriteExistingOutputFile [](start = 53, length = 27)

Why not "Force"? That's pretty conventional for meaning "overwrite existing file".

File: src/Sarif.Converters/FileInfoFactory.cs [hunk 7]
```diff
@@ -9,13 +9,15 @@ namespace Microsoft.CodeAnalysis.Sarif.Converters
...
     {
-        private readonly Dictionary<string, FileData> _fileInfoDictionary;
+        private readonly LoggingOptions _loggingOptions;
         private readonly Func<string, string> _mimeTypeClassifier;
+        private readonly Dictionary<string, FileData> _fileInfoDictionary;
 
-        internal FileInfoFactory(Func<string, string> mimeTypeClassifier)
-        {
+        internal FileInfoFactory(Func<string, string> mimeTypeClassifier, LoggingOptions loggingOptions)
+        {
```
Comment (lgolding) [inline src/Sarif.Converters/FileInfoFactory.cs]: >             [](start = 9, length = 12)

Remove trailing whitespace.

File: src/Sarif.Converters/FileInfoFactory.cs [hunk 8]
```diff
@@ -89,15 +91,17 @@ private void AddFile(PhysicalLocation physicalLocation)
                 filePath = uri.LocalPath;
             }
 
+            FileData fileData = FileData.Create(
```
Comment (lgolding) [inline src/Sarif.Converters/FileInfoFactory.cs]: >FileData [](start = 12, length = 8)

We should consider whether `Foo.Create` is sufficiently explicit to allow us to use `var` here.

File: src/Sarif/Core/FileData.cs [hunk 9]
```diff
@@ -15,6 +15,8 @@ namespace Microsoft.CodeAnalysis.Sarif
     /// </summary>
     public partial class FileData : ISarifNode
     {
+        internal static IFileSystem _fileSystem = new FileSystem();
```
Comment (lgolding) [inline src/Sarif/Core/FileData.cs]: >internal static IFileSystem _fileSystem = new FileSystem() [](start = 8, length = 58)

I suggest that you use constructor injection, as follows (well, it's not exactly "constructor" injection, but close enough):
```
public static FileData Create(TheArgs theArgs)
{
    Create(theArgs, new FileSystem());
}

// For testing only:
internal static FileData Create(TheArgs theArgs, IFileSystem fileSystem)
{
    // the real code.
}
```

File: src/Sarif/Core/FileData.cs [hunk 10]
```diff
@@ -25,7 +27,6 @@ public static FileData Create(Uri uri, SarifWriters.LoggingOptions loggingOption
             var fileData = new FileData()
             {
                 MimeType = mimeType,
-                Uri = uri
```
Comment (lgolding) [inline src/Sarif/Core/FileData.cs]: >Uri = uri [](start = 16, length = 9)

Why is this removed?

File: src/Sarif.Converters/ToolFormatConverter.cs [hunk 11]
```diff
@@ -60,44 +60,19 @@ public class ToolFormatConverter
...
             }
         }
 
-        /// <summary>Converts a tool log file to the SARIF format.</summary>
-        /// <param name="toolFormat">The tool format of the input file.</param>
-        /// <param name="inputFileName">The input log file name.</param>
-        /// <param name="outputFileName">The name of the file to which the resulting SARIF log shall be
-        /// written. This cannot be a directory.</param>
-        /// <remarks>This overload should be used only from test code.</remarks>
-        public void ConvertToStandardFormat(
```
Comment (lgolding) [inline src/Sarif.Converters/ToolFormatConverter.cs]: >ConvertToStandardFormat [](start = 20, length = 23)

Yay for deletion

File: BuildAndTest.cmd [hunk 12]
```diff
@@ -81,6 +81,15 @@ goto ExitFailed
 @REM Run all tests
 SET PASSED=true
 
+mstest /testContainer:bld\bin\Sarif.Converters.UnitTests\AnyCPU_%Configuration%\Sarif.Converters.UnitTests.dll
```
Comment (lgolding) [inline BuildAndTest.cmd]: >mstest [](start = 0, length = 6)

We were never running these before?? Good catch.

File: src/Sarif/IFileSystem.cs (hunk 1 above)
Comment (michaelcfanning) [reply src/Sarif/IFileSystem.cs]: we have a true compromise. no one is happy. everyone has agreed to drop doc comments against internal/private items. i have agreed to maintain/provide them for public.


---
In reply to: [136619704](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136619704) [](ancestors = 136619704,136618958,136618018)

File: src/ConvertToSarif/Program.cs (hunk 6 above)
Comment (michaelcfanning) [reply src/ConvertToSarif/Program.cs]: 'Force' could be a star wars reference, that's why. 'OverwriteExistingOutputFile' is perfectly descriptive. Think of it as an XML doc comment that's burned RIGHT INTO THE NAME, hee,

---
In reply to: [136626110](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136626110) [](ancestors = 136626110)

File: BuildAndTest.cmd (hunk 12 above)
Comment (michaelcfanning) [reply BuildAndTest.cmd]: Yep. Neglected this change when we refactored the converters.

---
In reply to: [136628913](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136628913) [](ancestors = 136628913)

File: src/Sarif.Converters/ToolFormatConverter.cs (hunk 11 above)
Comment (michaelcfanning) [reply src/Sarif.Converters/ToolFormatConverter.cs]: Yay! Death to unnecessary overload

---
In reply to: [136628767](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136628767) [](ancestors = 136628767)

File: src/Sarif/Core/FileData.cs (hunk 10 above)
Comment (michaelcfanning) [reply src/Sarif/Core/FileData.cs]: Per other comment, this is the wrong data for this member (which should be a relative uri only for nested items)

---
In reply to: [136628150](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136628150) [](ancestors = 136628150)

File: src/Sarif/Core/FileData.cs (hunk 9 above)
Comment (michaelcfanning) [reply src/Sarif/Core/FileData.cs]: why should I instantiate the file system for each call? Also, you approach will require mucking with every API that utilizes the file system object. my approach modifies a common internal member that arbitrary API could use.

---
In reply to: [136627741](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136627741) [](ancestors = 136627741)

File: src/Sarif.Converters/FileInfoFactory.cs (hunk 8 above)
Comment (michaelcfanning) [reply src/Sarif.Converters/FileInfoFactory.cs]: i don't care about having 'var' here. The value of 'var' in my opinion is 1) reduce line column count, 2) allows declarations to line up nicely in multi-var decls.

in this case, we have a single variable and FileData is not an egregiously long type name

---
In reply to: [136626728](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136626728) [](ancestors = 136626728)

File: src/Sarif/Core/FileData.cs (hunk 10 above)
Comment (lgolding) [reply src/Sarif/Core/FileData.cs]: Actually, no:
----
5.15.2:
If the file object represents a top-level file, then the uri property may be present. If present, it shall be equal to the name of the property within run.files (§5.12.9) whose value is this file object. If absent, it shall be interpreted as having that same value.

If the file object represents a nested file whose location relative to its parent can be expressed only by means of a path, then the uri property shall be present, and its value shall be a valid relative URI expressing that path.
----
In other words, for a top-level file, you are allowed to repeat the key as the `uri` property, but you don't have to. So, you can certainly justify this as "not needed, saves space". But it's not the case that this is always a "nested" URI.

---
In reply to: [136632876](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136632876) [](ancestors = 136632876,136628150)

File: src/Sarif/Core/FileData.cs (hunk 9 above)
Comment (lgolding) [reply src/Sarif/Core/FileData.cs]: Because global data is evil. You could imagine getting a test wrong by neglecting-to-set-up-a-new-mock-and-assign-it-to-the-static; then your test would accidentally use the mock from some previous test -- and who could tell which one? -- and your test would fail unpredictably.

With injection, you guarantee each test to use its own mock.

---
In reply to: [136633228](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136633228) [](ancestors = 136633228,136627741)

File: src/Sarif.UnitTests/Core/FileDataTests.cs (hunk 2 above)
Comment (lgolding) [reply src/Sarif.UnitTests/Core/FileDataTests.cs]: Not necessarily. See my other comment.

---
In reply to: [136618490](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136618490) [](ancestors = 136618490)

File: src/Sarif/Core/FileData.cs (hunk 9 above)
Comment (lgolding) [reply src/Sarif/Core/FileData.cs]: As we discussed, better pattern is:
```
public static FileData Create(TheArgs theArgs, IFileSystem fileSystem = null)
{
    if (fileSystem == null) fileSystem = new FileSystem();
}
```

---
In reply to: [136634455](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136634455) [](ancestors = 136634455,136633228,136627741)

Comment (michaelcfanning) [reply src/Sarif/Core/FileData.cs]: beautiful, like it

---
In reply to: [136639877](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136639877) [](ancestors = 136639877,136634455,136633228,136627741)

File: src/Sarif/Core/FileData.cs (hunk 10 above)
Comment (michaelcfanning) [reply src/Sarif/Core/FileData.cs]: Oops. ok, will resolve

---
In reply to: [136633850](https://github.com/Microsoft/sarif-sdk/pull/614#discussion_r136633850) [](ancestors = 136633850,136632876,136628150)
